File: Python/ExampleTests/common/simple_class_string.py
from . import tokens
DNC_STRING = '?DNC?'


class SimpleClassString:
    def __init__(self, an_int: str = '', a_string: str = ''):
        self.an_int = an_int
        self.a_string = a_string

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 2, 'SimpleClass')
        return cls(
            parts[0],
            parts[1]
        )

    def __str__(self):
        return (tokens.token(self.an_int) + ' ' +
                tokens.token(self.a_string))

    def _key(self):
        return (self.an_int, self.a_string)

    def __eq__(self, other):
        if not isinstance(other, SimpleClassString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'SimpleClassString({self})'

File: Swift/ExampleTests/common/SimpleClassString.swift
public struct SimpleClassString: CustomStringConvertible, Equatable {
    public let anInt: String
    public let aString: String

    public init(anInt: String, aString: String) {
        self.anInt = anInt
        self.aString = aString
    }

    public init(fromArray v: [String]) {
        self.anInt = v.count > 0 ? v[0] : ""
        self.aString = v.count > 1 ? v[1] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> SimpleClassString {
        let parts = Tokens.require(text, 2, "SimpleClass")
        return SimpleClassString(anInt: parts[0], aString: parts[1])
    }

    public var description: String {
        return Tokens.token(anInt) + " " + Tokens.token(aString)
    }

    public static let dncString = "?DNC?"

    public static func == (a: SimpleClassString, b: SimpleClassString) -> Bool {
        return (a.anInt == dncString || b.anInt == dncString || a.anInt == b.anInt)
            && (a.aString == dncString || b.aString == dncString || a.aString == b.aString)
    }
}
